File: app/src/main/java/com/kmaryin/varioussolutions/matrixmultiplying/java/MatrixMultiplier.java
package com.kmaryin.varioussolutions.matrixmultiplying.java;

/**
 * Created by konstantinmaryin on 2019-06-17.
 * All rights reserved
 */
public class MatrixMultiplier {
    private static final int MAX_NUMBER = 10;
    private static final int THREADS_COUNT = 5;

    private final int[][] m1;
    private final int[][] m2;
    private final int[][] result;
    private final int[][] asyncResult;

    private final int l;
    private final int m;
    private final int n;

    public MatrixMultiplier(final int l, final int m, final int n) {
        this.l = l;
        this.m = m;
        this.n = n;

        m1 = new int[l][m];
        m2 = new int[m][n];
        result = new int[l][n];
        asyncResult = new int[l][n];

        initMatrix(m1);
        initMatrix(m2);
    }

    private void initMatrix(int[][] m) {
        int rows = m.length;
        for (int i = 0; i < rows; ++i) {
            int cols = m[i].length;
            for (int j = 0; j < cols; ++j) {
                double nextRandom = Math.random();
                m[i][j] = (int) (nextRandom * MAX_NUMBER);
            }
        }
    }

    public void multiply() {
        for (int i = 0; i < l; ++i) {
            for (int j = 0; j < n; ++j) {
                int res = 0;

                for (int k = 0; k < m; ++k) {
                    res += m1[i][k] * m2[k][j];
                }

                result[i][j] = res;
            }
        }
    }

    public void multiplyAsync() {
        new AsyncMultiplyingExecutor(m1, m2, asyncResult, l, m, n).multiply(THREADS_COUNT);
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        appendMatrixToStringBuilder(sb, m1);
        sb.append('\n').append('\n');

        appendMatrixToStringBuilder(sb, m2);
        sb.append('\n').append('\n');

        sb.append("\nSync:\n");
        appendMatrixToStringBuilder(sb, result);
        sb.append('\n').append('\n');

        sb.append("\nAsync:\n");
        appendMatrixToStringBuilder(sb, asyncResult);

        return sb.toString();
    }

    private void appendMatrixToStringBuilder(StringBuilder sb, int[][] m) {
        int rows = m.length;
        for (int i = 0; i < rows; ++i) {
            int cols = m[i].length;
            for (int j = 0; j < cols; ++j) {
                sb.append(m[i][j]);
                if (j < cols - 1) {
                    sb.append(' ');
                }
            }
            if (i < rows - 1) {
                sb.append('\n');
            }
        }
    }
}

File: app/src/main/java/com/kmaryin/varioussolutions/matrixmultiplying/kotlin/MatrixMultiplier.kt
package com.kmaryin.varioussolutions.matrixmultiplying.kotlin

/**
 * Created by konstantinmaryin on 2019-06-17.
 * All rights reserved
 */
class MatrixMultiplier(private val l: Int, private val m: Int, private val n: Int) {
    private val m1: Array<IntArray> = Array(l) { IntArray(m) }
    private val m2: Array<IntArray> = Array(m) { IntArray(n) }
    private val result: Array<IntArray> = Array(l) { IntArray(n) }
    private val asyncResult: Array<IntArray> = Array(l) { IntArray(n) }

    init {
        initMatrix(m1)
        initMatrix(m2)
    }

    private fun initMatrix(m: Array<IntArray>) {
        for (i in 0 until m.size) {
            for (j in 0 until m[i].size) {
                val nextRandom = Math.random()
                m[i][j] = (nextRandom * MaxNumber).toInt()
            }
        }
    }

    fun multiply() {
        for (i in 0 until l) {
            for (j in 0 until n) {
                var res = 0

                for (k in 0 until m) {
                    res += m1[i][k] * m2[k][j]
                }

                result[i][j] = res
            }
        }
    }

    fun multiplyAsync() {
        AsyncMultiplierExecutor(m1, m2, asyncResult, l, m, n).multiply(ThreadsCount)
    }

    override fun toString(): String {
        val sb = StringBuilder()
        appendMatrixToStringBuilder(sb, m1)
        sb.append('\n').append('\n')

        appendMatrixToStringBuilder(sb, m2)
        sb.append('\n').append('\n')

        sb.append("\nSync:\n")
        appendMatrixToStringBuilder(sb, result)
        sb.append('\n').append('\n')

        sb.append("\nAsync:\n")
        appendMatrixToStringBuilder(sb, asyncResult)

        return sb.toString()
    }

    private fun appendMatrixToStringBuilder(sb: StringBuilder, m: Array<IntArray>) {
        for (i in 0 until m.size) {
            for (j in 0 until m[i].size) {
                sb.append(m[i][j])

                if (j < m[i].size - 1) {
                    sb.append(' ')
                }
            }

            if (i < m.size - 1) {
                sb.append('\n')
            }
        }
    }

    companion object {
        private const val MaxNumber = 10
        private const val ThreadsCount = 5
    }
}